src/transformStateWithClones.js

```diff
@@ -8,6 +8,32 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let previousState = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(previousState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const remove of action.keysToRemove) {
+          delete previousState[remove];
+        }
+        break;
+
+      case 'clear':
+        previousState = {};
```

Comment:
Assigning a new object to 'previousState' in the 'clear' case is correct, but consider renaming 'previousState' to 'stateCopy' for consistency and clarity.